wpilibc/src/main/native/cpp/experimental/controller/PIDController.cpp

```diff
@@ -0,0 +1,236 @@
+  {
+    std::lock_guard<wpi::mutex> lock(m_thisMutex);
+    m_minimumInput = minimumInput;
+    m_maximumInput = maximumInput;
+    m_inputRange = maximumInput - minimumInput;
+  }
+
+  SetReference(GetReference());
+}
+
+void PIDController::SetOutputRange(double minimumOutput, double maximumOutput) {
+  std::lock_guard<wpi::mutex> lock(m_thisMutex);
+  m_minimumOutput = minimumOutput;
+  m_maximumOutput = maximumOutput;
+}
+
+void PIDController::SetAbsoluteTolerance(double tolerance,
+                                         double deltaTolerance) {
+  std::lock_guard<wpi::mutex> lock(m_thisMutex);
+  m_toleranceType = Tolerance::kAbsolute;
+  m_tolerance = tolerance;
+  m_deltaTolerance = deltaTolerance;
+}
+
+void PIDController::SetPercentTolerance(double tolerance,
+                                        double deltaTolerance) {
+  std::lock_guard<wpi::mutex> lock(m_thisMutex);
+  m_toleranceType = Tolerance::kPercent;
+  m_tolerance = tolerance;
+  m_deltaTolerance = deltaTolerance;
+}
+
+double PIDController::GetError() const {
+  std::lock_guard<wpi::mutex> lock(m_thisMutex);
+  return GetContinuousError(m_currError);
+}
+
+/**
+ * Returns the change in error per second of the PIDController.
+ *
+ * @return The change in error per second.
+ */
+double PIDController::GetDeltaError() const {
+  std::lock_guard<wpi::mutex> lock(m_thisMutex);
+  return (m_currError - m_prevError) / GetPeriod();
+}
+
+double PIDController::Calculate(double measurement) {
+  // Storage for function inputs
+  double Kp;
+  double Ki;
+  double Kd;
+  double minimumOutput;
+  double maximumOutput;
+
+  // Storage for function input-outputs
+  double totalError;
+
+  {
+    std::lock_guard<wpi::mutex> lock(m_thisMutex);
+
+    Kp = m_Kp;
+    Ki = m_Ki;
+    Kd = m_Kd;
+    minimumOutput = m_minimumOutput;
+    maximumOutput = m_maximumOutput;
+
+    m_prevError = m_currError;
+    m_currError = GetContinuousError(m_reference - measurement);
+    totalError = m_totalError;
+  }
+
+  if (Ki != 0) {
+    totalError = clamp(totalError + m_currError * GetPeriod(),
+                       minimumOutput / Ki, maximumOutput / Ki);
+  }
+
+  double output = clamp(Kp * m_currError + Ki * totalError +
+                            Kd * (m_currError - m_prevError) / GetPeriod(),
+                        minimumOutput, maximumOutput);
+
+  {
+    std::lock_guard<wpi::mutex> lock(m_thisMutex);
+    m_totalError = totalError;
+    m_output = output;
+  }
+
+  return output;
+}
+
+double PIDController::Calculate(double measurement, double reference) {
+  SetReference(reference);
+  return Calculate(measurement);
+}
+
+void PIDController::Reset() {
+  std::lock_guard<wpi::mutex> lock(m_thisMutex);
+  m_prevError = 0;
+  m_totalError = 0;
+  m_output = 0;
+}
+
+void PIDController::InitSendable(SendableBuilder& builder) {
+  builder.SetSmartDashboardType("PIDController");
+  builder.SetSafeState([this]() { Reset(); });
+  builder.AddDoubleProperty("Kp", [this] { return GetP(); },
```

Comment:
These property names do not align with the existing PIDController sendable format - should be "p", "i", "d", "setpoint"